Auto-config regional cluster from labels and annotations, generating serviceSpec, PromxyServerGroup, and GrafanaDatasource

denis-ryzhkov opened the pull request:
* Part of https://github.com/k0rdent/kof/issues/77
* Similar to https://github.com/k0rdent/kof/pull/130 but for **regional** clusters.
* `kof-operator`:
  * Detects creation of a `ClusterDeployment` with `.metadata.labels...kof-cluster-role: regional`.
  * Reads the regional domain from its `.spec.config.clusterAnnotations`.
  * Creates related `PromxyServerGroup` and `GrafanaDatasource`.
* Regional `MultiClusterService`:
  * Detects creation of a `Cluster` with `.metadata.labels...kof-cluster-role: regional`.
  * Configures endpoints, cert email address, and other values - from `.Cluster.metadata.annotations`.
* Realized we can use Sveltos `policyRefs` to create `PromxyServerGroup` and `GrafanaDatasource` with less code and complexity - will try this in the next PR.
* Tested with **latest** `kcm` supporting `clusterAnnotations`:
```
kubectl logs -n kof deploy/kof-mothership-kof-operator

  INFO	Created PromxyServerGroup	{...
  "ClusterDeployment": {"name":"dryzhkov-aws-standalone-regional","namespace":"kcm-system"},
  "promxyServerGroupName": "dryzhkov-aws-standalone-regional-metrics"}

  INFO	Created GrafanaDatasource	{...
  "ClusterDeployment": {"name":"dryzhkov-aws-standalone-regional","namespace":"kcm-system"},
  "grafanaDatasourceName": "dryzhkov-aws-standalone-regional-logs"}

kubectl get PromxyServerGroup -n kof dryzhkov-aws-standalone-regional-metrics -o yaml

  apiVersion: kof.k0rdent.mirantis.com/v1alpha1
  kind: PromxyServerGroup
  metadata:
    creationTimestamp: "2025-03-19T14:05:11Z"
    generation: 1
    labels:
      app.kubernetes.io/managed-by: kof-operator
      k0rdent.mirantis.com/promxy-secret-name: kof-mothership-promxy-config
    name: dryzhkov-aws-standalone-regional-metrics
    namespace: kof
    resourceVersion: "103334"
    uid: REDACTED
  spec:
    cluster_name: dryzhkov-aws-standalone-regional
    http_client:
      basic_auth:
        credentials_secret_name: storage-vmuser-credentials
        password_key: password
        username_key: username
      dial_timeout: 5s
      tls_config:
        insecure_skip_verify: true
    path_prefix: /vm/select/0/prometheus/
    scheme: https
    targets:
    - vmauth.dryzhkov-aws-standalone-regional.REDACTED:443

kubectl get GrafanaDatasource -n kof dryzhkov-aws-standalone-regional-logs -o yaml

  apiVersion: grafana.integreatly.org/v1beta1
  kind: GrafanaDatasource
  metadata:
    creationTimestamp: "2025-03-19T14:05:11Z"
    generation: 1
    labels:
      app.kubernetes.io/managed-by: kof-operator
    name: dryzhkov-aws-standalone-regional-logs
    namespace: kof
    resourceVersion: "109639"
    uid: REDACTED
  spec:
    datasource:
      access: proxy
      basicAuth: true
      basicAuthUser: ${username}
      isDefault: false
      name: dryzhkov-aws-standalone-regional
      secureJsonData:
        basicAuthPassword: ${password}
      type: victoriametrics-logs-datasource
      url: https://vmauth.dryzhkov-aws-standalone-regional.REDACTED/vls
    instanceSelector:
      matchLabels:
        dashboards: grafana
    resyncPeriod: 5m0s
    valuesFrom:
    - targetPath: basicAuthUser
      valueFrom:
        secretKeyRef:
          key: username
          name: storage-vmuser-credentials
    - targetPath: secureJsonData.basicAuthPassword
      valueFrom:
        secretKeyRef:
          key: password
          name: storage-vmuser-credentials
  status:
    hash: REDACTED
    lastResync: "2025-03-19T14:15:12Z"
    uid: REDACTED
  
kubectl get clustersummaries -A -o wide

  NAMESPACE    NAME                                                                                         AGE   HELMCHARTS    KUSTOMIZEREFS   POLICYREFS
  kcm-system   kof-aws-dns-secrets-capi-dryzhkov-aws-standalone-regional                                    27m                                 Provisioned
  kcm-system   kof-child-cluster-capi-dryzhkov-aws-standalone-child                                         14m   Provisioned                   
  kcm-system   kof-regional-cluster-capi-dryzhkov-aws-standalone-regional                                   27m   Provisioned                   
  kcm-system   kof-storage-secrets-capi-dryzhkov-aws-standalone-child                                       14m                                 Provisioned
  kcm-system   kof-storage-secrets-capi-dryzhkov-aws-standalone-regional                                    27m                                 Provisioned
  kcm-system   p--dryzhkov-aws-standalone-child-capi-dryzhkov-aws-standalone-child                          14m                                 Provisioned
  kcm-system   p--dryzhkov-aws-standalone-regional-capi-dryzhkov-aws-standalone-regional                    27m                                 Provisioned

KUBECONFIG=dev/regional-kubeconfig helm get values -n kof kof-storage

  USER-SUPPLIED VALUES:
  cert-manager:
    email: denisr@denisr.com
  external-dns:
    enabled: true
  global:
    storageClass: ""
  grafana:
    ingress:
      host: grafana.dryzhkov-aws-standalone-regional.REDACTED
    security:
      credentials_secret_name: grafana-admin-credentials
  jaeger:
    ingress:
      enabled: true
      host: jaeger.dryzhkov-aws-standalone-regional.REDACTED
  tmp: ""
  victoria-logs-single:
    server:
      persistentVolume:
        storageClassName: ""
  victoriametrics:
    security:
      credentials_secret_name: storage-vmuser-credentials
      password_key: password
      username_key: username
    vmauth:
      ingress:
        enabled: true
        host: vmauth.dryzhkov-aws-standalone-regional.REDACTED
```
* End-to-end test to see the data in Grafana: OK.

gmlexx commented on charts/kof-mothership/templates/sveltos/regional-multi-cluster-service.yaml:
```diff
@@ -0,0 +1,62 @@
+
+      - name: kof-storage
+        namespace: {{ .Release.Namespace }}
+        template: kof-storage-{{ .Values.kcm.kof.charts.storage.version | replace "." "-" }}
```
We probably can change it to this, as all charts having the same version?
```suggestion
        template: kof-storage-{{ .Chart.Version | replace "." "-" }}
```

gmlexx commented on docs/DEV.md:
```diff
@@ -8,8 +8,10 @@
 cd kcm
 
+# TODO: Delete this workaround after the merge with local repo fix:
 # Downgrade Sveltos to avoid `server gave HTTP response to HTTPS client` for `kcm-local-registry`:
-yq -i '.dependencies[0].version = "0.45.0"' templates/provider/projectsveltos/Chart.yaml
+yq -i '.dependencies[0].version = "0.48.1"' templates/provider/projectsveltos/Chart.yaml
+yq -i '.dependencies[1].version = "0.45.0"' templates/provider/projectsveltos/Chart.yaml
```
Can we delete this now?

gmlexx commented on kof-operator/Dockerfile:
```diff
@@ -15,6 +17,7 @@ RUN go mod download
 COPY cmd/main.go cmd/main.go
 COPY api/ api/
 COPY internal/controller/ internal/controller/
+COPY pkg/ pkg/
```
Let's try to avoid this.
Here is the project that references grafana-operator successfully

https://github.com/grafana/tempo-operator/blob/main/go.mod

denis-ryzhkov replied:
Thanks, done: https://github.com/k0rdent/kof/pull/143/commits/5b93cf5a1159a6fa3af8ea5090a510210b75310c

gmlexx reviewed (changes requested):
Please check this version, it doesn't require all that cherry-picking files from operator
https://github.com/grafana/tempo-operator/blob/main/go.mod

Pay attention to replace section to make the import work
```
replace github.com/openshift/api => github.com/openshift/api v0.0.0-20230223193310-d964c7a58d75
```

denis-ryzhkov commented:
Ready for review round 2:
* Resolved git conflicts.
* Added optional `storageClass` from annotations to istio-regional.
* DRY-ed `repo_chart_name` in kof-mothership child-cluster-profile similar to kof-istio.
* Fixed post-merge finding https://github.com/k0rdent/kof/pull/142#pullrequestreview-2702945973
* Tested end-to-end, OK.

gmlexx commented on kof-operator/internal/controller/clusterdeployment_kof_cluster_role.go:
```diff
@@ -342,3 +328,132 @@ func locationIsTheSame(cloud string, c1, c2 *ClusterDeploymentConfig) bool {
+		Spec: kofv1alpha1.PromxyServerGroupSpec{
+			ClusterName: regionalClusterName,
+			Targets:     []string{fmt.Sprintf("vmauth.%s:443", regionalDomain)},
+			PathPrefix:  "/vm/select/0/prometheus/",
```
We need to parametrize this and set values for 3 different cases:

- Default vm-metrics enpoints
- Istio endpoints
- Custom endpoints

denis-ryzhkov replied:
Done: https://github.com/k0rdent/kof/pull/143/commits/9c3dfd7fbf6835cd5dc67c4b1cd6d6458984c17c

gmlexx commented on kof-operator/internal/controller/clusterdeployment_kof_cluster_role.go:
```diff
@@ -342,3 +328,132 @@ func locationIsTheSame(cloud string, c1, c2 *ClusterDeploymentConfig) bool {
+			Datasource: &grafanav1beta1.GrafanaDatasourceInternal{
+				Name:           regionalClusterName,
+				Type:           "victoriametrics-logs-datasource",
+				URL:            fmt.Sprintf("https://vmauth.%s/vls", regionalDomain),
```
We need to parametrize this and set values for 3 different cases:

Default vm-metrics enpoints
Istio endpoints
Custom endpoints

denis-ryzhkov replied:
Done: https://github.com/k0rdent/kof/pull/143/commits/9c3dfd7fbf6835cd5dc67c4b1cd6d6458984c17c

denis-ryzhkov commented on demo/cluster/aws-eks-regional.yaml:
```diff
@@ -7,114 +7,21 @@ metadata:
+      k0rdent.mirantis.com/kof-cert-email: mail@example.com
+      # Optional custom endpoints:
+      # k0rdent.mirantis.com/kof-logs-endpoint: https://vmauth.aws-eks-ue2.kof.example.com/vls
+      # k0rdent.mirantis.com/kof-metrics-endpoint: https://vmauth.aws-eks-ue2.kof.example.com/vm/select/0/prometheus
```
Looks like we should replace 2 annotations:
```
kof-metrics-endpoint:       https://vmauth.aws-ue2.kof.example.com/vm/select/0/prometheus
kof-logs-endpoint:          https://vmauth.aws-ue2.kof.example.com/vls
```
with 5 annotations:
```
kof-write-metrics-endpoint: https://vmauth.aws-ue2.kof.example.com/vm/insert/0/prometheus/api/v1/write
kof-read-metrics-endpoint:  https://vmauth.aws-ue2.kof.example.com/vm/select/0/prometheus
kof-write-logs-endpoint:    https://vmauth.aws-ue2.kof.example.com/vls/insert/opentelemetry/v1/logs
kof-read-logs-endpoint:     https://vmauth.aws-ue2.kof.example.com/vls
kof-write-traces-endpoint:  https://jaeger.aws-ue2.kof.example.com/collector
```

If we try to just cut off the `/path`:
```
kof-metrics-endpoint:       https://vmauth.aws-ue2.kof.example.com
kof-logs-endpoint:          https://vmauth.aws-ue2.kof.example.com
```
and auto-add the paths like `/vm/select/0/prometheus`
it may fail when these services are exposed like `/custom/location/select/0/prometheus`

Updating...

denis-ryzhkov replied:
* Done: https://github.com/k0rdent/kof/pull/143/commits/c9d75bc71a9fce7ab50a0e3a96e30a4a8af0feb1
* Not sure: https://github.com/k0rdent/kof/pull/143/commits/1efe0cbc856d23553e48a51e2b2b9620a89d411f

Will test end-to-end at AWS after review, on Monday.

gmlexx commented on kof-operator/internal/controller/clusterdeployment_kof_cluster_role.go:
```diff
@@ -342,3 +323,224 @@ func locationIsTheSame(cloud string, c1, c2 *ClusterDeploymentConfig) bool {
+	}
+	if !isIstio {
+		httpClient := promxyServerGroup.Spec.HttpClient
+		httpClient.TLSConfig = kofv1alpha1.TLSConfig{InsecureSkipVerify: true}
```
Why do we set InsecureSkipVerify? It's non-secure and should work fine without it, as we have a real letencrypt certificates.

denis-ryzhkov replied:
It was like this initially in the yaml. Right, should work without it. Done: https://github.com/k0rdent/kof/pull/143/commits/932136f578f76a4b5589388b5fcbec9bea20baa0

gmlexx commented on kof-operator/internal/controller/clusterdeployment_kof_cluster_role.go:
```diff
@@ -17,67 +23,82 @@ import (
+	WriteTracesAnnotation:  "https://jaeger.%s/collector",
+}
+var istioEndpoints = map[string]string{
+	ReadLogsAnnotation:    "http://%s-logs:9428/insert/opentelemetry/v1/logs",
```
```suggestion
	ReadLogsAnnotation:    "http://%s-logs:9428",
```

gmlexx reviewed (changes requested):
Tested this branch, the only error found is the incorrect logs read endpoint for Istio.
